### PR TITLE
Candidate page

### PR DESCRIPTION
**Pull request by adborden**
Phew! This was a PITA. I think I underestimated the work involved in merging the forks. Hopefully I've learned most of the "gotchas" so whatever is next will be a lighter lift but I'm a little weary about how much more I'll be able to get done this week.

This took a heck of a lot longer than I was hoping for so I ended up taking some shortcuts. I originally wanted to integrate the components better, like keep a consistent single code-style and swap out shared components like the money bar. Because of the effort involved, I decided not to munge the code too much and copy it over as-is. It's not ideal to have two different code-styles in the repo, or duplicate components, but since we're focusing on getting features out rather than a perfect codebase, I think this is the right approach.

I put all the ported modules into a `fork` module. Hopefully the name doesn't invoke any bad feelings, I couldn't think of anything more descriptive. By having these modules in their own module, it keeps  everything within the module consistent in code-style and architechture, rather than sprinkling these within the disclosure-frontend modules. If we're at the point we want to start integrating, we can pull modules one by one, convert the style and move it out of the fork module.

One more note on the branch names, I'm merging into `next` and treating that branch like "master". The plan is that once we're ready, we'll merge `next` into upstream's `master`.

screenshot from 2016-09-18 21-06-18

### CODE REVIEW

**Comment by adborden** [comment]
cc @kyleW

File: app/components/common/fork/candidate/index.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+      controllerAs: '$ctrl',
+      bindToController: {
+        candidate: '=',
+        opposing: '=',
+        supporting: '='
+      }
+    };
+  });
+
+function CandidatePageController ($rootScope, pageTitle) {
```
Comment (KyleW) [inline app/components/common/fork/candidate/index.js]: I believe $rootScope isn't being used.

File: app/components/common/fork/common/url.filter.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  .filter('url', function (base_url) {
+    return urlFilter;
+
+    function urlFilter (path, query) {
+      query = query || {};
+
+      // If relative path, prefix it with base_url
+      var url = /^https?:\/\//.test(path) ? path : base_url + '/#!' + path;
+
+      var querystring = Object.keys(query).map(function (param) {
```
Comment (KyleW) [inline app/components/common/fork/common/url.filter.js]: I believe $httpParamSerializer is what you're looking for here( https://docs.angularjs.org/api/ng/service/$httpParamSerializer)

**Review by KyleW — COMMENTED** [review]
Ugh, I don't know. This gets a candidates page up and running, but it doesn't do much to integrate with what's there.

It's going to make developing confusing since you'll constantly be trying to figure out which part of the repo you're in. And it's not going to get a consistent user experience. It's probably close enough, but you'll see the differences given that this page will be side by side with the other stuff.

But I'm eager to move on. If approving this unblocks development, then I guess I'm in favor of merging it.

**Comment by KyleW** [comment]
Wait, is this all the porting that needs to be done or is there more coming?

**Comment by adborden** [comment]
Thanks for the review! Were you able to build the branch yourself?

> It's going to make developing confusing since you'll constantly be trying to figure out which part of the repo you're in.

Yes, but is it that much different? Are you concerned with the code-style or the directory structure, or something else? For code-style I think we can make up a rule of thumb to be consistent within the file you're editing (or module?) and choose a single style later. For directory structure, maybe it's better if I just dump all the modules in `app/components/common` and ditch the `fork` module. That's the existing pattern in fact. Is that directory change cool with you?

> And it's not going to get a consistent user experience. It's probably close enough, but you'll see the differences given that this page will be side by side with the other stuff.

I don't think it's too inconsistent because I imported the base CSS from `disclosure-frontend` before starting. There are definitely inconsistencies with the mockup, but I figure we'll fix those as bugs in the near future.

> Wait, is this all the porting that needs to be done or is there more coming?

Yes, there is more coming, but not too much. This was the bulk of it, and there are some we wont' need. Here's a list of all the modules from odca-oakland. I'll do the referendum page next, then the home page.

| Done | Module | Needed? |
| --- | --- | --- |
| :white_check_mark: | ballot_item_detail | No |
| :white_check_mark: | candidate | Yes |
| :white_check_mark: | candidates | Yes |
| :white_check_mark: | committee | No |
| :white_check_mark: | common | Yes |
|  | four04 | Maybe |
|  | home.page | No, but some styles maybe |
| :white_check_mark: | locality.page | No |
| :white_check_mark: | money | Yes |
|  | referendum | Yes, for the supporting/opposing components |
| :white_check_mark: | static | Yes |
| :white_check_mark: | utils | No |

File: app/components/common/fork/candidate/index.js (hunk 1 above)
Comment (adborden) [reply app/components/common/fork/candidate/index.js]: Thanks! Fixed.

File: app/components/common/fork/common/url.filter.js (hunk 2 above)
Comment (adborden) [reply app/components/common/fork/common/url.filter.js]: Ah sweet! Good tip.

**Comment by KyleW** [comment]
Yeah, I think moving this into app/components/common/money and app/components/common/candidatedPage makes a lot more sense. At the least, it let's us avoid app/components/common/fork/common/index.js app/components/common/fork/common/fork.less which are totally redundant.

A few other thoughts:
- Do we really need the default module? Couldn't you just say {{ foo || 47 }} instead of {{foo | default:47}}? 
- The module names for "money" and "default" and "static_api" are too generic. At the very least, they should be prefixed with odca. And money should have some indication it's a bar chart and not a currency filter or something. But we can leave that as an issue somewhere and change it later.

Let's just get this merged asap and file issues for the rest.

**Comment by adborden** [comment]
Great I'll get this merged tonight. Here's what I'm hearing as outstanding:
- [ ] Replace `default` filter with simple boolean operators
- [ ] Remove `fork`, move modules to `app/components/common/`
- [ ] Prefix the modules and their components' names with `odca`
- [ ] Rename `moneyBar` to `moneyBarChart`

On the prefix, are you talking about the module name, or the names of the individual filters/factories/services?

**Comment by KyleW** [comment]
Yeah, that seems like the right list. Thanks for making those changes. And
I was mostly thinking the directive and service names just so its clear
that it's not a 3rd party library.

**Pull request merged**